liquidctl/cli.py

```diff
@@ -191,6 +213,30 @@ def _list_devices(devices, using_filters=False, device_id=None, verbose=False, d
     assert 'device' not in opts or len(devices) <= 1, 'too many results listed with --device'
 
 
+def _dev_status_obj(dev, status):
+    # unlike the humans, machines want to know everything; don't suppress
+    # devices without status data (typically from initialize)
+    if not status:
+        status = []
+
+    return {
+        'bus': dev.bus,
+        'address': dev.address,
+        # suppress the experimental suffix (check .experimental from list)
+        'description': dev.description.replace(' (experimental)', ''),
+        # JSON values can only be objects, arrays, numbers, strings, or
+        # false/null/true; convert all non-numeric status values to strings
+        'status': [
+            [
+                key,
+                val if isinstance(val, Number) else str(val),
+                unit
+            ]
```

Comment:
I for-see some issues with this, specificly using a list of (key, value, units) because some statuses do not have units. 
ie. my graphics card:
`{"bus": "i2c-4", "address": "0x2a", "description": "ASUS Strix RTX 2080 Ti OC", "status": [["Mode", "Off", ""]]}` 

And the temperature units are written in `°C` by leaving it like this it could lead to some issues if units change or parsing the non-ascii character `°`.

Imo this will lead to the encouragement of status parsing based off of array indexes rather than by status message type (for lack of a better phrase). 

I know that changing this would also require changing all of the device drivers and it may not be work it for an initial version of this. It might be worth changing unit to a type and then using that type to map to a string for the unit. 

type = 'temp' => `°C` or `°F` (configurable later?)
type = 'duty' => `%` 
type = 'speed' => `rpm`
type = 'color' => `''`
type = 'light_mode' => `''`

and then the json list would be a list of:
```json
{
    "label" : key,
    "value": val,
    "type": type,
    "unit": unit
}
```

Reply:
> And the temperature units are written in °C by leaving it like this it could lead to some issues if units change [...]

The units are as stable as the values or keys themselves.  That is, considerably stable, although it's something we can occasionally change if there's a compelling reason to do so.

> [...] or parsing the non-ascii character °.

That's a very fringe case, but nonetheless, it's supported: `LANG=C` will escape any non-ASCII characters.

Additionally, supporting JSON implies supporting Unicode:

> All Unicode characters may be placed within the quotation marks, except for the characters that must be escaped [...]
> —_[RFC 7159](https://tools.ietf.org/html/rfc7159.html#section-7)

> Imo this will lead to the encouragement of status parsing based off of array indexes rather than by status message type (for lack of a better phrase).

Unfortunately this already happens with the API, but it's caused by the external list, not by the tuple items themselves.

I think we can/should offer better documentation, with examples of recommended use.

Otherwise, the only way to have this not happen is to return something that cannot be indexed, like an object (of status items).  But that's silly, inefficient, and also raises another issue: it could be interpreted to mean that fields that exist today will always be there.  Yes, we could document that's not the case, but at that point we can just document good ways to use the list.

> It might be worth changing unit to a type and then using that type to map to a string for the unit.

This is related to some statues not having units: we simply don't know what type they are right now, and even with a new `get_status_v2` API, it would still be pretty hard to enumerate all types of statuses in a way that's useful to everyone.

Additionally, the units are, IMO, always necessary, and they already convey the type of the data items that are more important and/or more likely to be used.  For generic processing, I think that temperatures and speeds (in rpm) are the main sought-after data, and those are trivial to filter using the units.

Other use cases would already need to assume a very specific device (e.g. your graphics card, that reports the lighting mode), and so could simply match the key/label.

> and then the json list would be a list of:

Using objects for each status item would be terribly inefficient, both in space, as well as during encoding and decoding.  

Given the uniform structure of every status item (even if, admittedly, it's not very detailed), I really think that a tuple, or the closest thing to it in JSON, are more suitable.

But I'll see if I can come up with something better, as well as do some proper benchmarks.

Reply:
Your response addresses all of the concerns that I had, and I did not think of the encoding / decoding offences. Im going change my stance to this is all good and mark this as resolved.